### PR TITLE
DEV-497 Merge changes from https://github.com/gered/views.honeysql 

### PR DESCRIPTION
**Pull request by akhudek**
Huge update that merges https://github.com/gered/views.honeysql with our changes. Apologies in advance for the size of this, there was no easy way to do it in increments.

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -1,32 +1,29 @@
...
+  :url           "https://github.com/kirasystems/views"
 
-  :url "https://github.com/kirasystems/views"
+  :license       {:name "MIT License"
+                  :url "http://opensource.org/licenses/MIT"}
 
-  :license {:name "MIT License"
-            :url "http://opensource.org/licenses/MIT"}
+  :repositories  [["releases" {:url "https://clojars.org/repo"
+                               :sign-releases false
```
Comment (jpgauthier) [inline project.clj]: What's that? It wasn't there before

File: src/views/core.clj [hunk 2]
```diff
@@ -2,138 +2,207 @@
...
+   the subscription is successful, the subscriber will be sent the initial
+   data for the view."
+  [view-system {:keys [view-id parameters] :as view-sig} subscriber-key context]
+  (if-let [view (get-in @view-system [:views view-id])]
+    (let [namespace (view-namespace @view-system view-sig subscriber-key context)
+          view-sig  (->view-sig namespace view-id parameters)]
+      (if (authorized-subscription? @view-system view-sig subscriber-key context)
+        (subscribe-and-send! view-system view view-sig subscriber-key)
+        (do
+          (trace "subscription not authorized" view-sig subscriber-key context)
```
Comment (jpgauthier) [inline src/views/core.clj]: Why is that a trace, shouldn't it be a warning/info or maybe a debug?

File: src/views/core.clj [hunk 3]
```diff
@@ -2,138 +2,207 @@
...
+      ; amounts of entries with no subscriptions. this could easily happen over time
+      ; naturally for applications with long uptimes.
+      (update :subscribed
+              (fn [subscribed]
+                (if (empty? (get subscribed subscriber-key))
+                  (dissoc subscribed subscriber-key)
+                  subscribed)))))
+
+(defn- clean-up-unneeded-hashes
+  "Remove hashes for view-sigs which do not have any unsubscribers."
```
Comment (jpgauthier) [inline src/views/core.clj]: unsubscribers?

File: src/views/core.clj [hunk 4]
```diff
@@ -2,138 +2,207 @@
...
+    (update view-system :hashes dissoc view-sig)
+    view-system))
 
 (defn unsubscribe!
-  [view-system namespace view-id parameters subscriber-key]
+  "Removes a subscription to a view identified by view-sig for a subscriber
+   identified by subscriber-key. Additional context info can be passed in,
+   which will be passed to the view-system's namespace-fn (if provided)."
+  [view-system {:keys [view-id parameters] :as view-sig} subscriber-key context]
+  (trace "unsubscribing from view" view-sig subscriber-key)
```
Comment (jpgauthier) [inline src/views/core.clj]: Again, why not debug?

File: src/views/core.clj [hunk 5]
```diff
@@ -2,138 +2,207 @@
...
-               (update-in [:subscribed subscriber-key] disj [namespace view-id parameters])
-               (remove-from-subscribers [namespace view-id parameters] subscriber-key)
-               (update-hash! [namespace view-id parameters] nil))))) ;; see note #2 in NOTES.md
+           (let [namespace (view-namespace vs view-sig subscriber-key context)
+                 view-sig  (->view-sig namespace view-id parameters)]
+             (-> vs
+                 (remove-from-subscribed view-sig subscriber-key)
+                 (remove-from-subscribers view-sig subscriber-key)
+                 (assoc-hash view-sig nil) ;; see note #2 in NOTES.md
+                 (clean-up-unneeded-hashes view-sig)))))
```
Comment (jpgauthier) [inline src/views/core.clj]: Why is it needed? Is it to keep the memory usage as low as possible?

File: README.md [hunk 6]
```diff
@@ -4,27 +4,720 @@
...
+```clj
+(defprotocol IView
+  (data [this namespace parameters])
+  (relevant? [this namespace parameters hints])
+  (id [this]))
+```
+
+`id` simply returns a unique identifier for this view. `data` returns a
+copy of the underlying data represented by this view. `relevant?` 
+determines if a collection of hints are relevant to the view and is
```
Comment (akhudek) [inline README.md]: I'm pretty sure *are* is correct. If it were one hint, then it would be *is*.

Comment (jpgauthier) [inline README.md]: The subject here is a "collection of hints", not "hints". To remove ambiguity, you can remove "a collection of" and keep the "are".

File: README.md [hunk 7]
```diff
@@ -4,27 +4,720 @@
...
+A list of `IView` instances. These are the views that can be 
+subscribed to. Views can also be added/replaced in the system after 
+initialization by calling `views.core/add-views!`.
+
+#### `:put-hints-fn`
+
+A function that typically will be used by the different views plugin 
+libraries providing view implementations (such as 
+[views.sql](https://github.com/gered/views.sql) or
+[views.honeysql](https://github.com/gered/views.honeysql)) to add
```
Comment (akhudek) [inline README.md]: Probably not until we update our honeysql repo. We don't actually use our repo in our system (we use a tweaked version that handles the firm database).

File: src/views/core.clj (hunk 2 above)
Comment (akhudek) [inline src/views/core.clj]: Not sure, maybe all the trace commands should actually be debug?

File: src/views/core.clj (hunk 5 above)
Comment (akhudek) [inline src/views/core.clj]: Yes, doubt it's going to be a big issue but I don't see a problem with leaving it.

**Pull request merged**